filestore.go

```diff
@@ -0,0 +1,282 @@
+//go:build experimental
+// +build experimental
+
+// This file declares Go bindings for experimental features in TileDB.
+// Experimental APIs to do not fall under the API compatibility guarantees and
+// might change between TileDB versions
+
+package tiledb
+
+/*
+   	#cgo LDFLAGS: -ltiledb
+   	#cgo linux LDFLAGS: -ldl
+	#include <tiledb/tiledb_experimental.h>
+	#include <tiledb/tiledb_serialization.h>
+	#include <stdlib.h>
+*/
+import "C"
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"runtime"
+	"unsafe"
+)
+
+// CreateFileSchemaFor returns a schema optimized for the storage of the named file.
+// An empty name returns a general schema suitable for any file.
+func CreateFileSchemaFor(tdbCtx *Context, name string) (*ArraySchema, error) {
+	var fileURI *C.char
+	if name != "" {
+		fileURI = C.CString(name)
+		defer C.free(unsafe.Pointer(fileURI))
+	}
+
+	arraySchema := ArraySchema{context: tdbCtx}
+	ret := C.tiledb_filestore_schema_create(tdbCtx.tiledbContext, fileURI, &arraySchema.tiledbArraySchema)
+	if ret != C.TILEDB_OK {
+		return nil, fmt.Errorf("Error creating schema: %s", tdbCtx.LastError())
+	}
+	runtime.SetFinalizer(&arraySchema, func(arraySchema *ArraySchema) {
+		arraySchema.Free()
+	})
+
+	return &arraySchema, nil
+}
+
+// CreateGeneralFileSchema returns a general schema for the storage of a file.
+func CreateGeneralFileSchema(tdbCtx *Context) (*ArraySchema, error) {
+	return CreateFileSchemaFor(tdbCtx, "")
+}
+
+// File represents a TileDB filestore file.
+type File struct {
+	tdbCtx   *Context // the tiledb context for all operations
+	arrayURI string   // the uri of the array
+}
+
+// NewFile returns a file. The arrayURI is the uri of the underlying TileDB array that contains the file (see Export) or
+// in which the file will be stored (see Import)
+func NewFile(tdbCtx *Context, arrayURI string) *File {
+	return &File{
+		tdbCtx:   tdbCtx,
+		arrayURI: arrayURI,
+	}
+}
+
+// Size returns the uncompressed size of the file.
+func (f *File) Size() (int64, error) {
+	uri := C.CString(f.arrayURI)
+	defer C.free(unsafe.Pointer(uri))
+
+	var size C.size_t
+	ret := C.tiledb_filestore_size(f.tdbCtx.tiledbContext, uri, &size)
+	if ret != C.TILEDB_OK {
+		return 0, fmt.Errorf("Error getting file size: %s", f.tdbCtx.LastError())
+	}
+
+	return int64(size), nil
+}
+
+// Export writes the contents of f to the named local file.
+func (f *File) Export(name string) error {
```

Comment:
maybe this should be fileName or path since it is a full uri? Name implies just the relative name, not the full uri